resource-manager/src/main/java/org/snomed/module/storage/ModuleStorageCoordinator.java

```diff
@@ -0,0 +1,602 @@
+            return false;
+        }
+
+        moduleMetadata.setEdition(edition);
+        return doUpdateModuleMetadata(codeSystem, moduleId, effectiveTime, moduleMetadata);
+    }
+
+    private boolean doUploadRelease(String codeSystem, String moduleId, String effectiveTime, File rf2Package) {
+        boolean invalidArguments = invalidArguments(codeSystem, moduleId, effectiveTime, rf2Package);
+        if (invalidArguments) {
+            LOGGER.trace("Cannot upload RF2 package; invalid arguments.");
+            return false;
+        }
+
+        // Check if metadata already exists
+        String metadataResourcePath = getMetadataResourcePath(writeDirectory, codeSystem, moduleId, effectiveTime);
+        boolean existingMetadata = resourceManager.doDoesObjectExist(metadataResourcePath);
+        if (existingMetadata) {
+            LOGGER.trace("Cannot upload RF2 package; metadata already exists.");
+            return false;
+        }
+
+        // Check if RF2 package already exists
+        String rf2PackageResourcePath = getPackageResourcePath(writeDirectory, codeSystem, moduleId, effectiveTime, rf2Package.getName());
+        boolean existingRF2Package = resourceManager.doDoesObjectExist(rf2PackageResourcePath);
+        if (existingRF2Package) {
+            LOGGER.trace("Cannot upload RF2 package; RF2 package already exists.");
+            return false;
+        }
+
+        // Build metadata object
+        Optional<ModuleMetadata> optionalModuleMetadata = this.generate(codeSystem, moduleId, effectiveTime, rf2Package);
+        if (optionalModuleMetadata.isEmpty()) {
+            LOGGER.trace("Cannot upload RF2 package; no metadata available.");
+            return false;
+        }
+
+        // Write metadata to local temporary file
+        File tmpMetadataFile = FileUtils.doCreateTempFile("metadata.json");
+        FileUtils.writeToFile(tmpMetadataFile, optionalModuleMetadata.get());
+
+        // Upload metadata
+        resourceManager.doWriteResource(metadataResourcePath, asFileInputStream(tmpMetadataFile));
+
+        // Check if metadata uploaded
+        boolean newMetadata = resourceManager.doDoesObjectExist(metadataResourcePath);
+        if (!newMetadata) {
+            LOGGER.trace("Cannot upload RF2 package; failed to upload metadata.");
+            return false;
+        }
+
+        // Upload RF2 package
+        resourceManager.doWriteResource(rf2PackageResourcePath, asFileInputStream(rf2Package));
+
+        // Check if RF2 package uploaded
+        boolean newRF2Package = resourceManager.doDoesObjectExist(rf2PackageResourcePath);
+        if (!newRF2Package) {
+            LOGGER.trace("Failed to upload RF2 package. Removing previously uploaded metadata.");
+            boolean deleteResource = resourceManager.doDeleteResource(metadataResourcePath);
+            if (!deleteResource) {
+                LOGGER.trace("Cannot delete metadata; manual clean up required.");
+            }
```

Comment:
@pgwilliams If the metadata has been uploaded (step 1) but the package fails (step 2), then try and delete the metadata from step 1. If this fails, a manual clean up will be required from the team.